compiler/ir/backend.jvm/src/org/jetbrains/kotlin/backend/jvm/lower/TypeOperatorLowering.kt

```diff
@@ -0,0 +1,194 @@
+
+    override fun lower(irFile: IrFile) {
+        irFile.transformChildrenVoid()
+    }
+
+    private fun IrExpression.transformVoid() = transform(this@TypeOperatorLowering, null)
+
+    private val IrType.isReifiedTypeParameter: Boolean
+        get() = classifierOrNull?.safeAs<IrTypeParameterSymbol>()?.owner?.isReified == true
+
+    private fun lowerInstanceOf(argument: IrExpression, type: IrType) = with(builder) {
+        when {
+            type.isReifiedTypeParameter ->
+                irIs(argument, type)
+            argument.type.isNullable() && type.isNullable() -> {
+                irLetS(argument) { valueSymbol ->
+                    irOrOr(
+                        irEqualsNull(irGet(valueSymbol.owner)),
+                        irIs(irGet(valueSymbol.owner), type.makeNotNull())
+                    )
+                }
+            }
+            else -> irIs(argument, type.makeNotNull())
+        }
+    }
+
+    private fun lowerCast(argument: IrExpression, type: IrType): IrExpression = when {
+        type.isReifiedTypeParameter ->
+            builder.irAs(argument, type)
+        argument.type.isNullable() && !type.isNullable() ->
+            with(builder) {
+                irLetS(argument) { valueSymbol ->
+                    irIfNull(
+                        type,
+                        irGet(valueSymbol.owner),
+                        irThrow(irCall(typeCastException).apply {
+                            putValueArgument(0, irString("null cannot be cast to non-null type ${type.render()}"))
+                        }),
+                        lowerCast(irGet(valueSymbol.owner), type.makeNullable())
+                    )
+                }
+            }
+        argument.type.isSubtypeOfClass(type.erasedUpperBound.symbol) ->
+            argument
+        else ->
+            builder.irAs(argument, type)
+    }
+
+    override fun visitTypeOperator(expression: IrTypeOperatorCall): IrExpression = with(builder) {
+        at(expression)
+        return when (expression.operator) {
+            IrTypeOperator.IMPLICIT_COERCION_TO_UNIT -> {
+                irComposite {
+                    +expression.argument.transformVoid()
+                    +IrGetFieldImpl(UNDEFINED_OFFSET, UNDEFINED_OFFSET, unitInstanceField.symbol, expression.type)
+                }
+            }
+
+            // There is no difference between IMPLICIT_CAST and IMPLICIT_INTEGER_COERCION on JVM_IR.
+            // Instead, this is handled in StackValue.coerce.
+            IrTypeOperator.IMPLICIT_INTEGER_COERCION -> {
+                irImplicitCast(expression.argument.transformVoid(), expression.typeOperand)
+            }
+
+            IrTypeOperator.CAST ->
+                lowerCast(expression.argument.transformVoid(), expression.typeOperand)
+
+            IrTypeOperator.SAFE_CAST -> {
+                irLetS(expression.argument.transformVoid(), IrStatementOrigin.SAFE_CALL) { valueSymbol ->
+                    irIfThenElse(
```

Comment:
Everything looks good, except minor impact of bytecode size of reified inline functions with 'as?' : +4 instuctions

Reply:
Yes, previously we would generate a `reifiedOperationMarker` with a kind argument indicating a safe call. Upon inlining this would then be turned into the code we generate here. If you'd prefer, I could change it back to the old behavior?

Reply:
Now that I think about it, this actually makes a difference when calling the code from Java.
```kotlin
inline fun <refified T>f(x: Any?) = x as? T
```
Right now, calling `f(null)` from Java fails with an exception. With these changes the code runs without exceptions. So it turns out that this isn't just a size issue, we actually need the old behavior for backwards compatibility.

Reply:
@sfs Note that inline functions with reified parameters shouldn't be called from Java, since we generate them with ACC_SYNTHETIC in the current backend (and with ACC_PRIVATE before 1.3.40), and IIRC in JVM IR too. So the issue with the behavior inside the function body is only relevant for access via reflection, which should fail. In any case, I think it'd be nice to keep the same bytecode here as in the current backend

Reply:
Totally agree with @udalov. Please note that ACC_SYNTHETIC flag now is missed in IR

Reply:
I've implemented the changes in question and added an additional test to check that we generate reified operation markers for safe casts to reified types.